keps/1337-reclaim-within-cohort-while-borrowing/README.md

```diff
@@ -0,0 +1,482 @@
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Flopping between preempting pair of workloads](#flopping-between-preempting-pair-of-workloads)
+- [Design Details](#design-details)
+  - [API](#api)
+  - [Implementation overview](#implementation-overview)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Beta](#beta)
+    - [GA](#ga)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Reuse PreemptionPolicy enum](#reuse-preemptionpolicy-enum)
+  - [Minimal API](#minimal-api)
+  - [ReclaimWithinCohortConfig](#reclaimwithincohortconfig)
+  - [Move ReclaimWithinCohort under ReclaimWithinCohortConfig](#move-reclaimwithincohort-under-reclaimwithincohortconfig)
+<!-- /toc -->
+
+## Summary
+
+This KEP introduces ClusterQueue-level API to support workload preemption
+to reclaim quota within cohort for a workload that requires borrowing.
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Currently, there is no support for preemption to reclaim quota within cohort
+for workloads which require borrowing. This is too constraining for some setups.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- support preemption to reclaim quota within cohort for workloads that require borrowing.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- allow preemption of workloads, from other ClusterQueues, which are not borrowing.
```

Comment:
Also preempt withinClusterQueue while borrowing is non-goal

Reply:
Done